### ISSUE TITLE
Updated sensor frequency is not used by lamp.nearby_device

### ISSUE THREAD

**Issue by d3vp** [BIDMCDigitalPsychiatry/LAMP-platform#787]
Hi,

I have a self-hosted lamp server and I am using the Android app for testing sensor data collection. 

I have updated frequency for `lamp.gps`, `lamp.accelerometer` and `lamp.nearby_device` sensors but the new frequency is not used by `lamp.nearby_device`. (Following instructions here: https://docs.lamp.digital/start_here/updating_frequency). GPS and Accelerometer sensors work correctly i.e. they are using updated frequency values.

Here is the mongodb record:
```
  {
    _id: 'yk7pxagsvxfzzgtbe7dd',
    _parent: 'y598j0shthd4ktbw13kh',
    timestamp: 1699982142009,
    spec: 'lamp.nearby_device',
    name: 'Nearby Device',
    settings: { frequency: 0.00166666666 },
    _deleted: false
  }
```

I have checked the Android app code on Github and it seems that a fixed value of frequency is being used for Wifi sensor.

Could you please confirm if this feature is missing an implementation for `lamp.nearby_device`? 
Also, will it be implemented anytime soon? Currently, the frequency is too high and we are getting lots of data for wifi sensor.

Thanks!

**Comment by anjumg03** [comment]
![image (4)](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/assets/115161321/30bcce20-9c61-40cf-b6e5-6a3c8f1610af)
We didn't have that requirement.@d3vp

**Comment by anjumg03** [comment]
Even though the frequency is not configurable ; we will increase the default frequency value in SDK and the updated SDK will be available soon. @d3vp

**Comment by avaidyam** [comment]
@anjumg03 Every Sensor must have a configurable `frequency` parameter and a configurable `cellular_upload` parameter. Please see #145. //cc @ertjlane @carlan1

**Comment by jijopulikkottil** [comment]
@avaidyam 
Yes, But we have not applied individual timer for all sensors. Currently we have applied it to gps, device motion, accelerometer and all new sensor kit data.
Some sensors(like HealthKit and steps) are collecting all data, some sensors collecting data only when it's triggered (call, activity), other sensors(near by device) are using default 5 minutes interval to collect data. 

So we can implement the `frequency` (parameter) for all the sensors which are using some default intervals to fetch. Please confirm.

**Comment by sijitg** [comment]
Android lamp.nearby_device frequency was too high and we reduced it to 5 minutes 
@carlan1 Can we release the library?

**Comment by ZCOEngineer** [comment]
@avaidyam we shall make this change in both iOS and Android.
@carlan1 we can release the library after making this change too in Android.

**Comment by jijopulikkottil** [comment]
@carlan1 As per our last discussion, we have consolidated all sensors with its data collection behaviour.
[ mindLamp_sensors_datacollection_behaviour.xlsx](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/files/13649503/mindLamp_sensors_datacollection_behaviour.xlsx)

**Comment by jijopulikkottil** [comment]
In nearby_device sensor, Frequency parameter is implemented and the the WiFi data is collecting as expected. But the blue_tooth sensor data was not collecting!. We are working on it.

**Comment by carlan1** [comment]
For sensors such as device_usage, which operate on intervals of a defined duration, is it possible to change the duration? For example, device_usage collects data approximately every 30 minutes and each data point is a report over that 30 minute span. Is it possible to change this duration from 30 minutes to 17 minutes? (This is a hypothetical. Please do not actually change it.)

**Comment by jijopulikkottil** [comment]
@carlan1 we couldn't find any API to manage their collecting behaviour. We can only fetch the collected data (after 24 hours) using from-date and to-date.

**Comment by jijopulikkottil** [comment]
for lamp.nearby_device sensor, the data scheme of android was different from iOS. Now we made same for both iOS and Android. It will be:

```
{
            "sensor": "lamp.nearby_device",
            "data": {
                "strength": 0,
                "type": "wifi",
                "name": "DJsWifi",
                "address": "28:ee:52:b8:7f:e"
            },
            "timestamp": 1702964839366
}
```

**Comment by jijopulikkottil** [comment]
iOS build 2013.12.20 has been submitted to TestFlight. Its available for appstore connect users (internal testers).

**Comment by anjumg03** [comment]
Android staging build has been submitted for Google's review . It will be available for Open testing after review.

**Comment by jijopulikkottil** [comment]
@carlan1 let us know if we need to create GitHub release to submit latest production pointed build to App Store/Play Store.

**Comment by carlan1** [comment]
We can do that after testing. Thank you

**Comment by carlan1** [comment]
Which sensors have been updated in this version? @jijopulikkottil 

Also:

> But we have not applied individual timer for all sensors. Currently we have applied it to gps, device motion, accelerometer and all new sensor kit data.

But you also say 

> we couldn't find any API to manage their collecting behaviour. We can only fetch the collected data (after 24 hours) using from-date and to-date.

I am confused here, for which sensorkit sensors can you apply a custom frequency/timer?

**Comment by avaidyam** [comment]
> Also in iOS, blood_pressure sensor data was collecting only the latest value every 5 minutes from HealthKit, we will fix it to collect all records available till time.

@jijopulikkottil Please ensure that **ONLY NEW EVENTS** uploaded and not every single record each time. 

> we couldn't find any API to manage their collecting behaviour. We can only fetch the collected data (after 24 hours) using from-date and to-date.

@jijopulikkottil Are there any other odd issues with these collection API we should document? 
@carlan1 Can you please document this? It does make sense that device usage is not per-minute configurable. 

> for lamp.nearby_device sensor, the data scheme of android was different from iOS. Now we made same for both iOS and Android. 

@jijopulikkottil Can we ensure all other sensors have the same data format? 
@carlan1 Can you update documentation for this?

**Comment by jijopulikkottil** [comment]
@carlan1 
The SensorKit and HealthKit data collection behaviour are same. For both, we are collecting the data available in the SensorKit (or from the HealthKit) queried using from-date and to-date.

But when we check, the ambientLightSensor sensor data are available in high frequency,  We may get 1000 records for 5 minute duration. So to prevent the high volume of data, we applied a filter. We will skip data of ambientLightSensor, if the timestamp duration is less than 1 second. (This is achieved by looping the records and checking the time difference between previous record and current record).
As we implemented a default frequency to ambientLightSensor, we done the same capability to all other sensors. But we applied a default filtering (of 1 second) only for ambientLightSensor.

**Comment by jijopulikkottil** [comment]
> Please ensure that ONLY NEW EVENTS uploaded and not every single record each time.
@avaidyam for healthkit, we are collecting all events [no duplicates].
